File: backend/app/security/replay_cache.py
import time
from typing import Dict

class ReplayCacheEngine:
    def __init__(self, max_capacity: int = 10000, ttl_seconds: int = 86400):
        self.max_capacity = max_capacity
        self.ttl_seconds = ttl_seconds
        self._cache: Dict[str, float] = {}

    def has(self, key: str) -> bool:
        if not key:
            return False
        timestamp = self._cache.get(key)
        if timestamp is None:
            return False
        
        if time.time() - timestamp > self.ttl_seconds:
            del self._cache[key]
            return False
            
        return True

    def add(self, key: str) -> None:
        if not key:
            return
            
        if len(self._cache) >= self.max_capacity:
            # Evict oldest key
            oldest_key = next(iter(self._cache))
            del self._cache[oldest_key]
            
        self._cache[key] = time.time()

    def clear(self) -> None:
        self._cache.clear()

replay_cache = ReplayCacheEngine()

File: mobile/src/services/security/ReplayCache.ts
/**
 * Replay Attack Protection Cache
 * Adheres to Task 3 (Replay Protection).
 * Tracks processed packet IDs and digital signature hashes to prevent double-forwarding and replay flooding.
 */
export class ReplayCacheEngine {
  private cache: Map<string, number> = new Map();
  private maxEntries: number;
  private ttlMs: number;

  constructor(maxEntries: number = 5000, ttlMs: number = 24 * 3600 * 1000) {
    this.maxEntries = maxEntries;
    this.ttlMs = ttlMs;
  }

  /**
   * Checks if packet ID or signature has been processed within the expiration window.
   */
  public has(identifier: string): boolean {
    if (!identifier) return false;

    const timestamp = this.cache.get(identifier);
    if (!timestamp) return false;

    // Check if entry expired
    if (Date.now() - timestamp > this.ttlMs) {
      this.cache.delete(identifier);
      return false;
    }

    return true;
  }

  /**
   * Adds packet ID or signature hash to replay protection cache.
   */
  public add(identifier: string): void {
    if (!identifier) return;

    // Evict oldest entries if cache capacity exceeded
    if (this.cache.size >= this.maxEntries) {
      const oldestKey = this.cache.keys().next().value;
      if (oldestKey) {
        this.cache.delete(oldestKey);
      }
    }

    this.cache.set(identifier, Date.now());
  }

  /**
   * Clears expired entries from memory.
   */
  public cleanup(): void {
    const now = Date.now();
    for (const [key, timestamp] of this.cache.entries()) {
      if (now - timestamp > this.ttlMs) {
        this.cache.delete(key);
      }
    }
  }

  /**
   * Gets total tracked entries.
   */
  public size(): number {
    return this.cache.size;
  }

  /**
   * Resets cache state.
   */
  public clear(): void {
    this.cache.clear();
  }
}

export const ReplayCache = new ReplayCacheEngine();
